src/types/network.rs

```diff
@@ -323,53 +343,129 @@ where
+    }
+
+    fn nodes_mut(&mut self) -> NodesMut<Self::Node> {
+        NodesMut {
+            // NOTE(adn) Not the most efficient iterator, but due to my
+            // limited Rust-fu, I couldn't find anything better.
+            range: {
+                let mut v = Vec::default();
+
+                for n in self.from_graph.node_weights_mut() {
+                    v.push(n);
+                }
+
+                v.into_iter()
+            },
+        }
+    }
+}
+
+impl<W> Graph for Network<W>
+where
+    W: Default + fmt::Display + Clone,
+{
+    type Node = Artifact<String>;
+    type Edge = Dependency<usize, W>;
+    type Weight = W;
+    type NodeData = ArtifactType;
+    type EdgeData = DependencyType<W>;
+
+    fn neighbors(&self, node_id: &Id<Self::Node>) -> EdgeRefs<Id<Self::Node>, Id<Self::Edge>> {
+        let mut neighbors = Vec::default();
 
         if let Some(nid) = self.node_ids.get(node_id) {
             for eref in self.from_graph.edges(*nid) {
-                neighbours.push(EdgeReference {
-                    source: self.from_graph[eref.source()].id(),
-                    target: self.from_graph[eref.target()].id(),
+                neighbors.push(oscoin_graph_api::EdgeRef {
+                    from: self.from_graph[eref.source()].id(),
+                    to: self.from_graph[eref.target()].id(),
                     id: self.from_graph[eref.id()].id(),
                 })
             }
         }
 
-        neighbours
+        neighbors
+    }
+
+    fn get_node(&self, id: &Id<Self::Node>) -> Option<&Self::Node> {
+        self.node_ids
+            .get(id)
+            .and_then(|nid| Some(&self.from_graph[*nid]))
+    }
+
+    fn get_edge(&self, id: &Id<Self::Edge>) -> Option<&Self::Edge> {
+        self.edge_ids
+            .get(id)
+            .and_then(|eid| Some(&self.from_graph[*eid]))
     }
 
     fn nodes(&self) -> Nodes<Self::Node> {
         Nodes {
-            range: 0..self.from_graph.node_count(),
-            to_node_id: Box::new(move |i| &self.from_graph[node_index(i)]),
+            range: self
+                .from_graph
+                .raw_nodes()
+                .iter()
+                .map(|n| &n.weight)
+                .collect::<Vec<&Self::Node>>()
+                .into_iter(),
         }
     }
 
-    fn nodes_mut(&mut self) -> NodesMut<Self::Node> {
-        NodesMut {
-            // NOTE(adn) Not the most efficient iterator, but due to my
-            // limited Rust-fu, I couldn't find anything better.
-            range: {
-                let mut v = Vec::default();
+    fn edges(&self, node: &Id<Self::Node>) -> Edges<Self::Edge> {
+        let mb_id = self.node_ids.get(node);
+        match mb_id {
+            None => Edges {
+                range: Vec::new().into_iter(),
+            },
+            Some(nid) => Edges {
+                range: self
+                    .from_graph
+                    .edges(*nid)
+                    .map(|e| e.weight())
```

Comment:
I find it a bit confusing, that there is `weight` for nodes (this is from the petgraph library, right?) and `get_weight()` or `weight()` for edges. Is there any way we can make this more consistent?

Reply:
Yeah, unfortunately `petgraph` treats the "metadata" as "weight", so this lead to all kind of confusion. I do wonder if we can create a little (evil? :trollface: ) `PetgraphAdaptor` trait which wraps this call into something like `data` for nodes and `weight` for edges. I don't know, I will have a play and let you know 😉